File: web/js/dataset.js
/* dataset.js — typed, indexed view over window.EM_DATA
   (port of enc_roller/data/loader.py + dataset.py). Ranges decoded through
   EM.parseRange with the die PINNED; tiling anomalies recorded as warnings. */
(function () {
  "use strict";
  const EM = (globalThis.EM = globalThis.EM || {});
  const CATCHALL = /roll on\s+(the\s+)?table\s*[RS]3/i;

  function tileCheck(rows, die, label, warnings) {
    const ordered = rows.slice().sort((a, b) =>
      a.roll_low - b.roll_low || a.roll_high - b.roll_high);
    let expect = 1;
    for (const r of ordered) {
      if (r.roll_low > expect) warnings.push(`${label}: gap at ${expect}-${r.roll_low - 1}`);
      else if (r.roll_low < expect) warnings.push(`${label}: overlap at ${r.roll_low}-${expect - 1}`);
      expect = Math.max(expect, r.roll_high + 1);
    }
    if (expect - 1 < die) warnings.push(`${label}: gap at ${expect}-${die} (tail)`);
  }

  function findRow(rows, roll) {
    // rows sorted asc by roll_low, non-overlapping -> binary search
    let lo = 0, hi = rows.length - 1, ans = -1;
    while (lo <= hi) {
      const mid = (lo + hi) >> 1;
      if (rows[mid].roll_low <= roll) { ans = mid; lo = mid + 1; } else hi = mid - 1;
    }
    if (ans < 0) return null;
    const r = rows[ans];
    return (r.roll_low <= roll && roll <= r.roll_high) ? r : null;
  }

  EM.Dataset = class Dataset {
    constructor(raw) {
      const warnings = [];
      this._warnings = warnings;

      // ---- items (d1000) grouped by table, sorted -----------------------
      this.itemsByTable = {};
      for (const it of raw.items) {
        (this.itemsByTable[it.table] || (this.itemsByTable[it.table] = [])).push(it);
      }
      for (const t of Object.keys(this.itemsByTable)) {
        this.itemsByTable[t].sort((a, b) => a.roll_low - b.roll_low);
        tileCheck(this.itemsByTable[t], 1000, "item table " + t, warnings);
      }

      // ---- master (die pinned d100, parse from roll string) -------------
      this.masterRows = raw.master.map((r) => {
        const [lo, hi] = EM.parseRange(r.roll, 100);
        return { roll: r.roll, roll_low: lo, roll_high: hi, name: r.name || "", target: r.target || "" };
      }).sort((a, b) => a.roll_low - b.roll_low);
      tileCheck(this.masterRows, 100, "master", warnings);

      // ---- mechanics R1/R2/S1/S2 ----------------------------------------
      const pinned = { R1: 1000, R2: 20, S1: 1000, S2: 20 };
      this.mech = {};
      for (const key of Object.keys(pinned)) {
        const die = pinned[key];
        this.mech[key] = (raw.mech[key] || []).map((r) => {
          const [lo, hi] = EM.parseRange(r.roll, die);
          const name = r.name || "";
          return {
            table: key, roll: r.roll, roll_low: lo, roll_high: hi, name: name,
            reroll: !!r.reroll, is_r3_catchall: CATCHALL.test(name),
            wpn_adj: r.wpn_adj || null,
          };
        }).sort((a, b) => a.roll_low - b.roll_low);
        tileCheck(this.mech[key], die, key, warnings);
      }

      // ---- artifact power tables ----------------------------------------
      this.powerByNum = {};
      for (const t of raw.powerTables) {
        const die = t.die | 0;
        const entries = t.entries.map((e) => {
          let [lo, hi] = EM.parseRange(e.roll, die);
          if (hi < lo) { warnings.push(`power ${t.num}: backwards range ${e.roll} -> swapped`); [lo, hi] = [hi, lo]; }
          return { roll: e.roll, roll_low: lo, roll_high: hi, text: e.text || "", name: e.text || "" };
        }).sort((a, b) => a.roll_low - b.roll_low);
        this.powerByNum[t.num] = { num: t.num, category: t.category || "", die: die, entries: entries, note: t.note || null };
        tileCheck(entries, die, "power " + t.num, warnings);
      }

      // ---- enhancement selector (d100) ----------------------------------
      const ed = raw.enhancement;
      const edie = ed.die | 0 || 100;
      this.enhancement = ed.entries.map((e) => {
        const [lo, hi] = EM.parseRange(e.roll, edie);
        return { roll: e.roll, roll_low: lo, roll_high: hi, type: e.type || "", name: e.type || "", reroll: !!e.reroll };
      }).sort((a, b) => a.roll_low - b.roll_low);
      tileCheck(this.enhancement, edie, "enhancement", warnings);
      this.enhancementDescriptions = ed.descriptions || [];

      // ---- search index (name+category+table+subcategory+page) ----------
      this._search = [];
      for (const t of Object.keys(this.itemsByTable)) {
        for (const it of this.itemsByTable[t]) {
          const hay = [it.name, it.category, it.table, it.subcategory || "",
            it.page == null ? "" : (Array.isArray(it.page) ? it.page.join(" ") : it.page)]
            .join(" ").toLowerCase();
          this._search.push([it, hay]);
        }
      }
    }

    get itemCount() { return this._search.length; }

    find(rows, roll) { return rows && rows.length ? findRow(rows, roll) : null; }

    findNearest(rows, roll) {
      if (!rows || !rows.length) return [null, "no rows"];
      let best = null, bestDist = null;
      for (const r of rows) {
        let dist;
        if (roll < r.roll_low) dist = r.roll_low - roll;
        else if (roll > r.roll_high) dist = roll - r.roll_high;
        else return [r, "exact"];
        if (bestDist === null || dist < bestDist) { best = r; bestDist = dist; }
      }
      return [best, `nearest range ${best.roll_low}-${best.roll_high}`];
    }

    search(tokens) {
      const toks = tokens.map((t) => t.toLowerCase()).filter((t) => t.length);
      if (!toks.length) return this._search.map((x) => x[0]);
      return this._search.filter(([, hay]) => toks.every((t) => hay.indexOf(t) !== -1)).map((x) => x[0]);
    }

    dataWarnings() { return this._warnings.slice(); }
  };
})();

File: desktop/enc_roller/data/dataset.py
"""dataset.py — indexed, read-only view over the loaded models.

All range lookups go through ``find`` / ``find_nearest`` (bisect over sorted,
non-overlapping ``(roll_low, roll_high)`` ranges) so no engine or UI site
re-implements matching.  ``search`` powers the Library's live filter.
"""

from __future__ import annotations

import bisect

from . import loader
from .models import EnhRow, Item, MasterRow, MechRow, PowerTable


class Dataset:
    """Everything the engine and UI need, indexed for lookup and search."""

    def __init__(self, loaded: dict):
        self.items_by_table: dict[str, list[Item]] = loaded["items_by_table"]
        self.master_rows: list[MasterRow] = loaded["master_rows"]
        self.mech: dict[str, list[MechRow]] = loaded["mech"]
        self.power_by_num: dict[str, PowerTable] = loaded["power_by_num"]
        self.enhancement: list[EnhRow] = loaded["enhancement"]
        self._warnings: list[str] = list(loaded.get("warnings", []))

        # Cache the sorted low-bounds per row-list for bisect (keyed by identity).
        self._lows_cache: dict[int, list[int]] = {}

        # Flat search index for the Library: (item, lowercased haystack).
        self._search_index: list[tuple[Item, str]] = []
        for rows in self.items_by_table.values():
            for it in rows:
                hay = " ".join(str(x) for x in (
                    it.name, it.category, it.table, it.subcategory or "",
                    it.page if it.page is not None else "",
                )).lower()
                self._search_index.append((it, hay))

    # ---- classmethod constructor ----------------------------------------- #
    @classmethod
    def from_app(cls, app_dir) -> "Dataset":
        return cls(loader.load_all(app_dir))

    # ---- range lookups ---------------------------------------------------- #
    def _lows(self, rows) -> list[int]:
        key = id(rows)
        lows = self._lows_cache.get(key)
        if lows is None:
            lows = [r.roll_low for r in rows]
            self._lows_cache[key] = lows
        return lows

    def find(self, rows, roll: int):
        """Return the row whose ``[roll_low, roll_high]`` contains ``roll``, else None.

        Assumes ``rows`` is sorted ascending by ``roll_low`` (loader guarantees it).
        """
        if not rows:
            return None
        lows = self._lows(rows)
        idx = bisect.bisect_right(lows, roll) - 1
        if idx < 0:
            return None
        r = rows[idx]
        if r.roll_low <= roll <= r.roll_high:
            return r
        return None

    def find_nearest(self, rows, roll: int):
        """For a runtime gap (R1/S1): return ``(row, note)`` for the nearest range.

        Defensive fallback only — the shipped data tiles fully, so this should
        never fire.  ``note`` describes the distance for the trace.
        """
        if not rows:
            return None, "no rows"
        best = None
        best_dist = None
        for r in rows:
            if roll < r.roll_low:
                dist = r.roll_low - roll
            elif roll > r.roll_high:
                dist = roll - r.roll_high
            else:
                return r, "exact"
            if best_dist is None or dist < best_dist:
                best, best_dist = r, dist
        return best, f"nearest range {best.roll_low}-{best.roll_high}"

    # ---- search ----------------------------------------------------------- #
    def search(self, tokens: list[str]) -> list[Item]:
        """AND of lowercased tokens across name/category/table/subcategory/page."""
        toks = [t.lower() for t in tokens if t.strip()]
        if not toks:
            return [it for it, _ in self._search_index]
        return [it for it, hay in self._search_index
                if all(t in hay for t in toks)]

    @property
    def item_count(self) -> int:
        return len(self._search_index)

    # ---- diagnostics ------------------------------------------------------ #
    def data_warnings(self) -> list[str]:
        return list(self._warnings)
